subsys/net/lib/download_client/src/http.c

```diff
@@ -130,31 +134,41 @@ static int http_header_parse(struct download_client *client, size_t *hdr_len)
 		client->buf[i] = tolower(client->buf[i]);
 	}
 
-	p = strstr(client->buf, "http/1.1 206");
-	if (!p) {
-		p = strstr(client->buf, "http/1.1 404");
-		if (p) {
-			LOG_ERR("Server response was 404: file not found");
-			return -1;
-		}
-		if (client->proto == IPPROTO_TLS_1_2
-		   || IS_ENABLED(CONFIG_DOWNLOAD_CLIENT_RANGE_REQUESTS)) {
-			LOG_ERR("Server did not honor partial content request");
-			return -1;
-		}
-		p = strstr(client->buf, "http/1.1 200");
-		if (!p) {
-			LOG_ERR("Server response is not 200 Success");
-			return -1;
+	/* Look for the status code just after "http/1.1 " */
+	p = strstr(client->buf, "http/1.1 ");
+	if (p == NULL) {
+		LOG_ERR("Server response missing HTTP/1.1");
+		return -1;
+	}
+	p += strlen("http/1.1 ");
+
+	const unsigned int http_status = strtoul(p, &q, 10);
+
+	if (q == NULL) {
+		LOG_ERR("Server response malformed: status code not found");
+		return -1;
+	}
+	if (http_status != expected_status) {
+		/* Truncate the server response at the first CR or LF after the
+		 * status and message so we can log it. Normally we can't
+		 * modify the response like this since we'd need to look at
+		 * data beyond this point, but we're already in a failing
+		 * condition.
+		 */
+		while ((*q != 0) && (*q != '\r') && (*q != '\n')) {
+			q++;
 		}
+		*q = 0;
```

Comment:
nit: use `\0` when comparing/assigning the NULL-termination char (instead of 0).
It's the same but it's clearer that you are terminating a string.